mrmustard/lab/gates.py

```diff
@@ -177,6 +178,47 @@ def __init__(
     def X_matrix(self):
         return gaussian.rotation_symplectic(self.angle.value)
 
+    def U(self, cutoffs: Sequence[int]):
+
+        angles = self._parse_modes_and_args(cutoffs)
+        num_modes = len(cutoffs)
+
+        # calculate rotation unitary for each mode and concatenate with outer product
+        Ur = None
+        for idx, cutoff in enumerate(cutoffs):
+            theta = math.arange(cutoff) * angles[idx]
+            if Ur is None:
+                Ur = math.diag(math.make_complex(math.cos(theta), math.sin(theta)))
+            else:
+                U_next = math.diag(math.make_complex(math.cos(theta), math.sin(theta)))
+                Ur = math.outer(Ur, U_next)
+
+        # return total unitary with indexes reordered according to MM convetion
+        return math.transpose(
+            Ur,
+            list(range(0, 2 * num_modes, 2)) + list(range(1, 2 * num_modes, 2)),
+        )
+
+    def _parse_modes_and_args(self, cutoffs):
+        num_modes = len(cutoffs)
+        modes = self.modes  # modes in which the gate is acting on
+        args = self.angle.value
+        num_args = (
+            args.shape[0] if len(args.shape.as_list()) > 0 else 1
+        )  # number or arguments given to the gate
+        angles = np.zeros((num_modes,))
+        if num_args == 1:
+            # one arg for all modes
+            angles[modes] = args
+        elif num_args == len(modes):
+            # an arg for each mode
+            angles[modes] = args
```

Comment:
Should this be angles = args? Otherwise there is no difference from the outcome of the preceding if statement

Reply:
yes! Thanks for spotting this